crates/bevy_render/src/view/visibility/render_groups.rs

```diff
@@ -0,0 +1,678 @@
+    /// as a slice of bitmasks.
+    pub fn bits(&self) -> &[u64] {
+        self.layers.as_slice()
+    }
+
+    /// Returns `true` if the internal bitmask is on the heap.
+    pub fn is_allocated(&self) -> bool {
+        self.layers.spilled()
+    }
+
+    fn layer_info(layer: usize) -> (usize, u64) {
+        let buffer_index = layer / 64;
+        let bit_index = layer % 64;
+        let bit = 1u64 << bit_index;
+
+        (buffer_index, bit)
+    }
+
+    fn extend_buffer(&mut self, other_len: usize) {
+        let new_size = std::cmp::max(self.layers.len(), other_len);
+        self.layers.reserve_exact(new_size - self.layers.len());
+        self.layers.resize(new_size, 0u64);
+    }
+
+    fn iter_layers(mut buffer: u64) -> impl Iterator<Item = RenderLayer> + 'static {
+        let mut layer: usize = 0;
+        std::iter::from_fn(move || {
+            if buffer == 0 {
+                return None;
+            }
+            let next = buffer.trailing_zeros() + 1;
+            buffer >>= next;
+            layer += next as usize;
+            Some(RenderLayer(layer - 1))
+        })
+    }
+}
+
+impl<T: Into<RenderLayer>> From<T> for RenderLayers {
+    fn from(layer: T) -> Self {
+        let mut layers = Self {
+            layers: SmallVec::default(),
+        };
+        layers.add(layer);
+        layers
+    }
+}
+
+impl<R: Into<RenderLayer>> FromIterator<R> for RenderLayers {
+    fn from_iter<T: IntoIterator<Item = R>>(i: T) -> Self {
+        i.into_iter().fold(Self::empty(), |mut mask, g| {
+            mask.add(g);
+            mask
+        })
+    }
+}
+
+impl Default for RenderLayers {
+    fn default() -> Self {
+        Self::from(DEFAULT_RENDER_LAYER)
+    }
+}
+
+/// Component on an entity that controls which cameras can see it.
+///
+/// There are two kinds of render groups:
+/// - [`RenderLayers`]: These are grouping categories that many cameras can view (see [`CameraView`]).
+/// - *Camera entity*: This is a specific camera that the entity is affiliated with. This is especially
+/// useful for UI in combination with [`PropagateRenderGroups`].
+///
+/// An entity can be a member of multiple [`RenderLayers`] in addition to having a camera affiliation.
+///
+/// ### Default behavior
+///
+/// A default-constructed `RenderGroups` will include [`DEFAULT_RENDER_LAYER`].
+/// If you don't want that, then use [`Self::empty`], [`Self::new_with_camera`], or
+/// [`Self::from::<RenderLayer>`].
+///
+/// ### Entity default behavior
+///
+/// All entities without a [`RenderGroups`] component are in [`DEFAULT_RENDER_LAYER`] by
+/// default (layer 0). If you add a [`RenderGroups`] component to an entity, it may no longer
+/// be in the default layer if the [`RenderGroups`] component doesn't include it.
+///
+/// For example, if you do `entity.insert(RenderGroups::from(RenderLayer(1)))`, then `entity`
+/// will only be in layer 1. You can instead do:
+/**
+```no-run
+// Option 1: default
```

Comment:
Might be worth adding a `Vec<usize>` constructor for convenience? Although forcing users through the default constructor rather than making assumptions about layer 0 might make sense.